river/metrics/cluster/sd_validation.py

```diff
@@ -0,0 +1,175 @@
+import math
+
+from river import stats, utils
+
+from . import base
+
+__all__ = ["SD"]
+
+
+class SD(base.InternalClusMetric):
+    """The SD validity index (SD).
+
+    The SD validity index (SD) [^1] is a more recent clustering validation measure. It is composed of
+    two terms:
+
+    * Scat(NC) stands for the scattering within clusters,
+
+    * Dis(NC) stands for the dispersion between clusters.
+
+    Like DB and SB, SD measures the compactness with variance of clustered objects and separation
+    with distnace between cluster centers, but uses them in a different way. The smaller the value
+    of SD, the better.
+
+    In the original formula for SD validation index, the ratio between the maximum and the actual
+    number of clusters is taken into account. However, due to the fact that metrics are update live
+    along with the data stream, this ratio will be automatically set to default as 1.
+
+    Examples
+    --------
+
+    >>> from river import cluster
+    >>> from river import stream
+    >>> from river import metrics
+
+    >>> X = [
+    ...     [1, 2],
+    ...     [1, 4],
+    ...     [1, 0],
+    ...     [4, 2],
+    ...     [4, 4],
+    ...     [4, 0],
+    ...     [-2, 2],
+    ...     [-2, 4],
+    ...     [-2, 0]
+    ... ]
+
+    >>> k_means = cluster.KMeans(n_clusters=3, halflife=0.4, sigma=3, seed=0)
+    >>> metric = metrics.cluster.SD()
+
+    >>> for x, _ in stream.iter_array(X):
+    ...     k_means = k_means.learn_one(x)
+    ...     y_pred = k_means.predict_one(x)
+    ...     metric = metric.update(x, y_pred, k_means.centers)
+
+    >>> metric
+    SD: 4.332702
+
+    References
+    ----------
+    [^1]: Halkidi, M., Vazirgiannis, M., & Batistakis, Y. (2000). Quality Scheme Assessment in the
+          Clustering Process. Principles Of Data Mining And Knowledge Discovery, 265-276.
+          DOI: 10.1007/3-540-45372-5_26
+
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._center_all_points = {}
+        self._overall_variance = {}
+        self._cluster_variance = {}
+        self._centers = {}
+        self._initialized = False
+
+    @staticmethod
+    def _calculate_dispersion_nc(centers):
+        min_distance_clusters = math.inf
+        max_distance_clusters = -math.inf
+        sum_inverse_distances = 0
+
+        n_clusters = len(centers)
+
+        for i in range(n_clusters):
+            for j in range(i + 1, n_clusters):
+                distance_ij = math.sqrt(
+                    utils.math.minkowski_distance(centers[i], centers[j], 2)
+                )
+                if distance_ij > max_distance_clusters:
+                    max_distance_clusters = distance_ij
+                if distance_ij < min_distance_clusters:
+                    min_distance_clusters = distance_ij
+                sum_inverse_distances += 1 / distance_ij
+
+        return max_distance_clusters / min_distance_clusters * sum_inverse_distances
```

Comment:
Are divisions by zero a concern here? (and possibly in other similar situations)

Reply:
Usually, since the number of centers passed through the function have a length different from `0`, the result would also be different from `0` (as I believe the user would be responsible enough :D). However, since a division by zero is also feasible, I would implement a `try - except` here and everywhere concerned.

Reply:
This seems to be a good compromise!